### PR TITLE
fix(browser_adapter.ts): baseElement.getAttribute

### LINKED ISSUE #3441: HTML5LocationStrategy doesn't work (even with base attr)
Hello, I'm getting this error when I tried to use HTML5LocationStrategy:

```
Error during instantiation of LocationStrategy! (App -> Router -> function (_platformStrategy, href) {
        var $__0 = this;
        this._platformStrategy = _platformStrategy;
        this._subject = new EventEmitter();
        var browserBaseHref = isPresent(href) ? href : this._platformStrategy.getBaseHref();
        if (isBlank(browserBaseHref)) {
          throw new BaseException("No base href set. Either provide a binding to \"appBaseHrefToken\" or add a base element.");
        }
        this._baseHref = stripTrailingSlash(stripIndexHtml(browserBaseHref));
        this._platformStrategy.onPopState((function(_) {
          return $__0._onPopState(_);
        }));
      } -> LocationStrategy). ORIGINAL ERROR: TypeError: baseElement.attr is not a function

 ORIGINAL STACK: TypeError: baseElement.attr is not a function
    at getBaseElementHref (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:20758:24)
    at BrowserDomAdapter.execute.BrowserDomAdapter.getBaseHref (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:21164:24)
    at new $__0 (https://code.angularjs.org/2.0.0-alpha.32/router.dev.js:491:32)
    at execute.ReflectionCapabilities.factory (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:8752:26)
    at Injector.execute.Injector._new (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27630:25)
    at InjectorDynamicStrategy.execute.InjectorDynamicStrategy.getObjByKeyId (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27502:48)
    at Injector.execute.Injector._getByKeyUnbounded (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27761:39)
    at Injector.execute.Injector._getByKey (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27715:27)
    at Injector.execute.Injector._getByDependency (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27703:27)
    at Injector.execute.Injector._new (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27601:68)

Error: null
    at InstantiationError.BaseException (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:7724:25)
    at InstantiationError.AbstractBindingError (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:9577:66)
    at new InstantiationError (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:9624:64)
    at Injector.execute.Injector._new (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27694:21)
    at InjectorDynamicStrategy.execute.InjectorDynamicStrategy.getObjByKeyId (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27502:48)
    at Injector.execute.Injector._getByKeyUnbounded (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27761:39)
    at Injector.execute.Injector._getByKey (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27715:27)
    at Injector.execute.Injector._getByDependency (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27703:27)
    at Injector.execute.Injector._new (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27601:68)
    at InjectorDynamicStrategy.execute.InjectorDynamicStrategy.getObjByKeyId (https://code.angularjs.org/2.0.0-alpha.32/angular2.dev.js:27502:48)
```

I already added the attribute base on the head of my HTML, which in other ticket solved a similar issue.

Here is the bootstrap of my app:

```js
import { App } from './app.ts';
import { bootstrap, bind, httpInjectables } from 'angular2/angular2';
import { routerInjectables, HTML5LocationStrategy, LocationStrategy, appBaseHrefToken } from 'angular2/router';
import { reflector } from 'angular2/src/reflection/reflection';
import { ReflectionCapabilities } from 'angular2/src/reflection/reflection_capabilities';

export function main() {
    reflector.reflectionCapabilities = new ReflectionCapabilities();

    // I tried to fix the issue with this line, but it didn't change anything
    bind(appBaseHrefToken).toValue('/');

    bootstrap(App, [
        routerInjectables,
        httpInjectables,
        bind(LocationStrategy).toClass(HTML5LocationStrategy)
    ]);
}
```

I think that the error could be in this line:
https://github.com/angular/angular/blob/70bc485755e4bc65c27c2fa8998618cb780ac931/modules/angular2/src/dom/browser_adapter.ts#L308

It shouldn't be `return baseElement.getAttribute('href');`?

My environment:
Angular 2.0.0-alpha.32.
Google Chrome 43.0.2357.132 (64-bit) on Fedora 22.
I only have the necessary third party libraries: traceur, systemjs and typescript.

Many Thanks!

### PR DESCRIPTION
**Pull request by PatrickJS**
currently throwing an error

``` error
Error during instantiation of LocationStrategy! (RouterLink -> Router -> Location -> LocationStrategy). 
ORIGINAL ERROR: TypeError: baseElement.attr is not a function
```

closes #3441

### CODE REVIEW

File: modules/angular2/src/dom/browser_adapter.ts [hunk 1]
```diff
@@ -287,7 +287,7 @@ function getBaseElementHref(): string {
       return null;
     }
   }
-  return baseElement.attr('href');
+  return baseElement.getAttribute('href');
```
Comment (tbosch) [inline modules/angular2/src/dom/browser_adapter.ts]: Hey, could you add a unit test for this?

File: modules/angular2/test/dom/dom_adapter_spec.ts [hunk 2]
```diff
@@ -49,5 +49,49 @@ export function main() {
...
+    it('should getBaseHref only relative url', () => {
+
+      var initialValue = 'bass';
+
+      var $base = DOM.createElement('base');
+      DOM.setAttribute($base, 'href', initialValue);
+      var $head = DOM.defaultDoc().head;
+      DOM.appendChild($head, $base);
+
+      var subject = DOM.getBaseHref();
```
Comment (gdi2290) [inline modules/angular2/test/dom/dom_adapter_spec.ts]: this fails due to [`baseElement being cache`](https://github.com/gdi2290/angular/blob/patch-1/modules/angular2/src/dom/browser_adapter.ts#L285) from the prevous test

File: modules/angular2/src/dom/browser_adapter.ts (hunk 1 above)
Comment (gdi2290) [inline modules/angular2/src/dom/browser_adapter.ts]: I added tests but ran into a problem with [baseElement cache](https://github.com/angular/angular/pull/3214/files#r35475994)

Comment (tbosch) [inline modules/angular2/src/dom/browser_adapter.ts]: Could you add a `resetBaseElement` method to the `DomAdapter` (and all implementations). You can narrow it down to only the JS/Dart adapter via `if (DOM.supportsDOMEvents()) {...}`

Comment (gdi2290) [inline modules/angular2/src/dom/browser_adapter.ts]: I think I did something wrong here when adding `resetBaseElement` and `getBaseHref` to the other adapters

**Comment by gdi2290** [comment]
I'm having trouble correctly testing this in Dart. Can anyone help?

**Comment by btford** [comment]
@gdi2290 sure– what's the problem?

**Comment by gdi2290** [comment]
@btford with Dart I'm trying not run these tests for Dart (since it's only in js atm).  Using `if (DOM.supportsDOMEvents)` [travis dart error](https://github.com/angular/angular/pull/3214/files#diff-b4743922539bf28ebff17607917d2620R53) doesn't seem to work when avoiding the Dart version. I recall an environment variable (`IS_DARTIUM`) that is used to narrow down a test to js so should I use that rather than `supportsDOMEvents` as Tobias suggested? Unless he meant for me to write out `getBaseHref` for Dart too (I know he is on vaca this week)

**Comment by btford** [comment]
> Unless he meant for me to write out getBaseHref for Dart too (I know he is on vaca this week)

I think that's what he meant, but I'm not sure. :/

**Comment by btford** [comment]
Moving this to alpha-35.

**Comment by gdi2290** [comment]
@btford can you take over this PR? you can either close and recreate it or push to my repo `git push patrickjs yourbranch:patch-1` for it to show up in this PR. All that's left is Dart support and I don't want this PR to live through another version :/

**Comment by btford** [comment]
Yes, I'll get this in today.

**Comment by btford** [comment]
Fixed and pushed to presubmit. :+1:

**Comment by gdi2290** [comment]
thanks! :+1:

**Pull request closed without merging**